### PR TITLE
Remove legacy VectorIndexState

### PR DESCRIPTION
**Pull request by madmax983**
Removed legacy `VectorIndexState` struct which was replaced by multi-property `vector_indexes` (DashMap). 
Refactored `CurrentStorage` to use the DashMap for all vector operations. 
Introduced `get_default_vector_property_name()` to maintain backward compatibility for legacy APIs by deterministically selecting the first index (alphabetical). 
This resolves the TODO in `src/storage/current.rs:185`.

---
*PR created automatically by Jules for task [9806428225326814239](https://jules.google.com/task/9806428225326814239) started by @madmax983*

### CODE REVIEW

File: src/storage/current.rs [hunk 1]
```diff
@@ -667,15 +626,16 @@ impl CurrentStorage {
...
+        if let Some(property_name) = self.get_default_vector_property_name() {
+            if let Some(entry) = self.vector_indexes.get(&property_name) {
+                return VectorIndexCheckpointData::enabled(
+                    property_name,
+                    entry.value().config.clone(),
+                );
+            }
         }
+
+        VectorIndexCheckpointData::disabled()
```
[the remark this answers is not in the document]
Comment (madmax983) [reply src/storage/current.rs]: @jules

File: src/storage/current.rs [hunk 2]
```diff
@@ -2427,8 +2401,12 @@ impl CurrentStorage {
...
     /// Used by the query planner for statistics.
     pub fn vector_count(&self) -> usize {
-        let state = self.vector_index_state.read();
-        state.index.as_ref().map(|idx| idx.len()).unwrap_or(0)
+        if let Some(prop_name) = self.get_default_vector_property_name() {
+            if let Some(entry) = self.vector_indexes.get(&prop_name) {
+                return entry.value().index.len();
+            }
+        }
+        0
```
[the remark this answers is not in the document]
Comment (madmax983) [reply src/storage/current.rs]: @jules

**Comment by madmax983** [comment]
> ## Code Review: Remove legacy VectorIndexState
> **Summary**: This PR successfully removes the legacy `VectorIndexState` struct and refactors `CurrentStorage` to use the multi-property `vector_indexes` DashMap exclusively. The implementation introduces `get_default_vector_property_name()` to maintain backward compatibility for legacy APIs.
> 
> ### ✅ Strengths
> 1. **Clean Removal**: All references to the legacy single-property state have been completely removed
> 2. **Backward Compatibility**: The new `get_default_vector_property_name()` method maintains deterministic behavior for legacy APIs by selecting the alphabetically first property name
> 3. **Code Simplification**: Removes ~22 lines of legacy code and eliminates dual maintenance burden
> 4. **Performance Optimization**: Single-property case is optimized (O(1) vs scanning all keys)
> 5. **Consistent Architecture**: All vector operations now go through the same DashMap path
> 
> ### 🔍 Code Quality
> **Positive:**
> 
> * Proper error handling maintained throughout
> * Lock usage eliminated where appropriate (removed unnecessary RwLock reads)
> * Comments clearly document backward compatibility intent
> * Method visibility appropriately set to private (`fn get_default_vector_property_name()`)
> 
> ### ⚠️ Concerns & Suggestions
> #### 1. **Potential Breaking Change in Multi-Property Scenarios**
> The deterministic selection strategy (alphabetically first) could cause subtle breaking changes:
> 
> ```rust
> // Old behavior: First enabled index becomes default (insertion order)
> db.enable_vector_index("image_embedding", config1)?;
> db.enable_vector_index("text_embedding", config2)?;
> // Legacy API would use "image_embedding"
> 
> // New behavior: Alphabetically first
> db.enable_vector_index("image_embedding", config1)?;
> db.enable_vector_index("text_embedding", config2)?;
> // Legacy API now uses "text_embedding" (different!)
> ```
> 
> **Impact**: Users with multiple vector properties who rely on legacy APIs (`find_similar`, `find_similar_by_embedding`, etc.) may get different results after this change.
> 
> **Recommendation**:
> 
> * Document this behavior change in the PR description or CHANGELOG
> * Consider adding a deprecation warning to legacy APIs
> * Add migration guidance for users with multi-property setups
> 
> #### 2. **Checkpoint Persistence Limitation**
> The `get_vector_index_config()` method only persists the "default" property:
> 
> ```rust
> pub fn get_vector_index_config(&self) -> VectorIndexCheckpointData {
>     if let Some(property_name) = self.get_default_vector_property_name() {
>         if let Some(entry) = self.vector_indexes.get(&property_name) {
>             return VectorIndexCheckpointData::enabled(property_name, entry.value().config.clone());
>         }
>     }
>     VectorIndexCheckpointData::disabled()
> }
> ```
> 
> **Issue**: With multi-property vector indexes, only one property configuration is checkpointed. This appears to be a pre-existing limitation, but it's now more visible.
> 
> **Questions**:
> 
> * Is multi-property checkpoint persistence planned? (Issue [Comprehensive Vector Index API: Multi-property support and unified DX #389](https://github.com/madmax983/GallifreyDB/issues/389)?)
> * Should this method be updated to checkpoint all properties?
> * Should there be a comment noting this limitation?
> 
> #### 3. **Missing Test Coverage**
> There are no tests specifically for `get_default_vector_property_name()` behavior:
> 
> **Suggested Tests**:
> 
> ```rust
> #[test]
> fn test_get_default_vector_property_name_empty() {
>     let storage = CurrentStorage::new();
>     assert!(storage.get_default_vector_property_name().is_none());
> }
> 
> #[test]
> fn test_get_default_vector_property_name_single() {
>     let storage = CurrentStorage::new();
>     storage.enable_vector_index("embedding", config)?;
>     assert_eq!(storage.get_default_vector_property_name(), Some("embedding".to_string()));
> }
> 
> #[test]
> fn test_get_default_vector_property_name_alphabetical() {
>     let storage = CurrentStorage::new();
>     storage.enable_vector_index("zebra", config1)?;
>     storage.enable_vector_index("apple", config2)?;
>     storage.enable_vector_index("middle", config3)?;
>     // Should return "apple" (alphabetically first)
>     assert_eq!(storage.get_default_vector_property_name(), Some("apple".to_string()));
> }
> ```
> 
> #### 4. **Minor: Clippy & Formatting**
> Before committing, ensure:
> 
> ```shell
> cargo clippy --all-targets --all-features -- -D warnings
> cargo fmt --all
> cargo test
> ```
> 
> Per [CLAUDE.md](https://github.com/madmax983/GallifreyDB/blob/trunk/CLAUDE.md#%EF%B8%8F-mandatory-pre-commit-quality-checks), these are mandatory pre-commit checks.
> 
> ### 🎯 Performance Considerations
> **Positive:**
> 
> * Eliminated RwLock contention on hot path (removed `vector_index_state.read()` calls)
> * Optimized single-property case (O(1) vs O(n))
> * DashMap provides better concurrent access patterns
> 
> **Neutral:**
> 
> * Multi-property case requires scanning all keys to find minimum (O(n)), but this only affects legacy APIs
> 
> ### 🔒 Security
> No security concerns identified. The change maintains proper ID validation and error handling.
> 
> ### 📊 Test Coverage
> **Existing tests** should cover most scenarios:
> 
> * `test_enable_vector_index` - Basic functionality
> * `test_double_enable_vector_index_fails` - Duplicate prevention
> * `test_multi_property_vector_index_node_deletion` - Multi-property scenarios
> * Multiple `find_similar*` tests - Legacy API coverage
> 
> **Recommendation**: Add explicit tests for `get_default_vector_property_name()` behavior with multiple properties to document the alphabetical selection strategy.
> 
> ### ✅ Approval Recommendation
> **APPROVE with suggestions for follow-up:**
> 
> This is a solid refactoring that removes technical debt and simplifies the codebase. The implementation is correct and maintains backward compatibility with a deterministic strategy.
> 
> **Follow-up items** (can be addressed in separate PRs):
> 
> 1. Add tests for `get_default_vector_property_name()` deterministic behavior
> 2. Document behavior change in CHANGELOG (alphabetical selection vs insertion order)
> 3. Consider deprecating legacy single-property APIs in favor of explicit property names
> 4. Evaluate multi-property checkpoint persistence support
> 
> **Note**: The alphabetical selection strategy is a subtle behavior change that users with multiple vector properties should be aware of. Consider adding a migration note if this is released in a version update.
> 
> Great work cleaning up this legacy code! The implementation is clean and the backward compatibility approach is reasonable. 🚀

@jules

**Comment by madmax983** [comment]
> ## Code Review: Remove legacy VectorIndexState
> ### Summary
> This PR successfully removes the legacy `VectorIndexState` struct and consolidates all vector indexing operations to use the multi-property `DashMap`-based approach. The refactoring is clean, well-tested, and maintains backward compatibility through the new `get_default_vector_property_name()` helper.
> 
> ### ✅ Strengths
> 1. **Clean Removal**: Successfully eliminates technical debt by removing the legacy `VectorIndexState` struct and `vector_index_state` field entirely.
> 2. **Deterministic Behavior**: The new `get_default_vector_property_name()` method provides deterministic selection (alphabetically first property) which is critical for backward compatibility and testability.
> 3. **Good Test Coverage**: Two new tests comprehensively verify:
>    
>    * Deterministic property selection across insertion orders (`test_legacy_default_property_selection_determinism`)
>    * Vector counting through the default property (`test_legacy_vector_count`)
> 4. **Performance Optimization**: The single-index fast path (`if len == 1`) avoids unnecessary iteration when only one vector index exists.
> 5. **Consistent Refactoring**: All methods now consistently use `get_default_vector_property_name()`:
>    
>    * `prepare_vector_search()` (line 1337-1343)
>    * `prepare_vector_search_raw()` (line 1838-1844)
>    * `get_indexed_property_name()` (line 570)
>    * `get_vector_property_name()` (line 2314)
>    * `vector_count()` (line 2395)
>    * `get_vector_index_checkpoint_data()` (line 625)
> 6. **Documentation**: Good inline comments explaining the determinism requirement and DashMap iteration order guarantees.
> 
> ### 💡 Suggestions
> #### 1. **Performance Consideration: Repeated Min Computation**
> The `get_default_vector_property_name()` method is called frequently and performs a full iteration + min() on every call when multiple indexes exist. Consider caching the result:
> 
> ```rust
> // In CurrentStorage struct
> cached_default_property: RwLock<Option<String>>
> 
> // In get_default_vector_property_name()
> fn get_default_vector_property_name(&self) -> Option<String> {
>     if self.vector_indexes.is_empty() {
>         return None;
>     }
>     
>     // Check cache first
>     let cache = self.cached_default_property.read();
>     if let Some(ref name) = *cache {
>         if self.vector_indexes.contains_key(name) {
>             return Some(name.clone());
>         }
>     }
>     drop(cache);
>     
>     // Compute and cache
>     let result = if self.vector_indexes.len() == 1 {
>         self.vector_indexes.iter().next().map(|r| r.key().clone())
>     } else {
>         self.vector_indexes.iter().map(|r| r.key().clone()).min()
>     };
>     
>     if let Some(ref name) = result {
>         *self.cached_default_property.write() = Some(name.clone());
>     }
>     
>     result
> }
> 
> // Invalidate cache in enable_vector_index()
> *self.cached_default_property.write() = None;
> ```
> 
> **Impact**: With many vector properties, the current O(n) scan on every query could add measurable overhead. Given that GallifreyDB targets <1µs single-hop queries, even a small overhead matters.
> 
> #### 2. **Code Duplication in Error Messages**
> Lines 1339-1343 and 1840-1844 have very similar error handling. Consider extracting to a helper:
> 
> ```rust
> fn get_vector_index_for_default_property(&self) -> Result<Ref<String, VectorIndexEntry>> {
>     let prop_name = self.get_default_vector_property_name().ok_or_else(|| {
>         crate::utils::error::Error::Vector(crate::utils::error::VectorError::IndexError(
>             "Vector index is not enabled".to_string(),
>         ))
>     })?;
>     
>     self.vector_indexes.get(&prop_name).ok_or_else(|| {
>         crate::utils::error::Error::Vector(crate::utils::error::VectorError::IndexError(
>             format!("Vector index not found for property '{}'", prop_name),
>         ))
>     })
> }
> ```
> 
> #### 3. **Test Coverage: Edge Case**
> Consider adding a test for the transition when the alphabetically-first index is removed:
> 
> ```rust
> #[test]
> fn test_default_property_updates_when_first_removed() {
>     let storage = CurrentStorage::new();
>     let config = HnswConfig::new(4, DistanceMetric::Cosine).with_capacity(100);
>     
>     storage.enable_vector_index("a_prop", config.clone()).unwrap();
>     storage.enable_vector_index("b_prop", config.clone()).unwrap();
>     
>     assert_eq!(storage.get_indexed_property_name(), Some("a_prop".to_string()));
>     
>     // TODO: Add disable_vector_index() method and test that default switches to "b_prop"
>     // This is important for backward compatibility when the default index is removed
> }
> ```
> 
> #### 4. **Documentation Clarity**
> The method `get_indexed_property_name()` at line 570 has a doc comment saying "Note: For multi-property setups, use [`list_vector_indexes`] instead." This is good, but it might be worth adding a warning about the alphabetical selection behavior since users might expect chronological (first-enabled) ordering.
> 
> ### 🔍 Security & Correctness
> ✅ **No security concerns identified**
> 
> * No unsafe code introduced
> * No new panics (no unwrap/expect calls)
> * Proper error handling maintained
> 
> ✅ **Correctness verified**
> 
> * DashMap iteration is correctly handled (non-deterministic order acknowledged)
> * Min operation on String keys works correctly (lexicographic ordering)
> * No race conditions introduced
> 
> ### 📊 Adherence to Guidelines
> ✅ **Coding Standards ([CODING_STANDARDS.md](https://github.com/madmax983/GallifreyDB/blob/trunk/docs/CODING_STANDARDS.md))**
> 
> * Type safety maintained (proper use of NodeId, etc.)
> * No unwrap() calls
> * Inline optimization hints preserved (`#[inline]`)
> 
> ✅ **Architecture ([ARCHITECTURE.md](https://github.com/madmax983/GallifreyDB/blob/trunk/docs/ARCHITECTURE.md))**
> 
> * Maintains hybrid storage principles
> * No performance regression expected for current state queries
> 
> ⚠️ **Testing ([TESTING.md](https://github.com/madmax983/GallifreyDB/blob/trunk/TESTING.md))**
> 
> * New tests added (good!)
> * Consider running `just coverage` to ensure 85%+ coverage maintained
> 
> ### 🎯 Recommendations
> **Before Merge:**
> 
> 1. Run `cargo clippy --all-targets --all-features -- -D warnings` to ensure no warnings
> 2. Run `cargo fmt --all` for consistent formatting
> 3. Run `just coverage-check` to verify coverage thresholds (≥85% line coverage)
> 
> **Future Enhancements:**
> 
> 1. Consider implementing the caching optimization if profiling shows `get_default_vector_property_name()` is a hot path
> 2. Add `disable_vector_index()` method to complete the API
> 3. Consider deprecating the "default property" concept entirely in favor of explicit property names in all APIs (breaking change for next major version)
> 
> ### ✅ Approval Status
> **This PR is well-architected and ready to merge** pending CI checks and the pre-commit quality verification mentioned above.
> 
> The refactoring successfully removes technical debt while maintaining backward compatibility through a well-designed deterministic selection strategy. The test coverage is solid, and the code quality meets GallifreyDB's high standards.
> 
> **Score: 9/10** - Excellent work! Minor performance optimization opportunity noted but not blocking.
> 
> _Reviewed by Claude Code following [GallifreyDB development guidelines](https://github.com/madmax983/GallifreyDB/blob/trunk/CLAUDE.md)_

@jules

**Comment by madmax983** [comment]
> ## Code Review: Remove legacy VectorIndexState
> ### Summary
> This PR successfully removes the legacy `VectorIndexState` struct and consolidates all vector index operations to use the multi-property `DashMap` (`vector_indexes`). The changes are well-implemented with good backward compatibility considerations.
> 
> ### ✅ **Strengths**
> 1. **Clean Refactoring**: The removal of the legacy `VectorIndexState` and `RwLock` simplifies the codebase by eliminating dual state management. All vector index operations now consistently use the `DashMap`.
> 2. **Deterministic Default Selection**: The new `get_default_vector_property_name()` method uses alphabetical ordering (`min()`) to deterministically select a default property for legacy APIs. This is a solid approach for backward compatibility.
> 3. **Performance Optimization**: The method includes an optimization for single-index scenarios (O(1) without sorting) and properly documents that DashMap iteration is unordered, requiring a full scan.
> 4. **Test Coverage**: Good addition of two new tests:
>    
>    * `test_legacy_default_property_selection_determinism`: Validates alphabetical selection
>    * `test_legacy_vector_count`: Validates count calculation via DashMap
> 5. **No Breaking Changes**: The refactor maintains all public APIs while cleaning up internal state management.
> 
> ### ⚠️ **Issues & Concerns**
> #### **1. Critical: Checkpoint Persistence Still Uses Legacy Single-Property Model**
> **Location**: `src/storage/current.rs:625-636`
> 
> The `snapshot_vector_index_config()` method still returns `VectorIndexCheckpointData` which is designed for **single-property** vector indexes:
> 
> ```rust
> pub fn snapshot_vector_index_config(
>     &self,
> ) -> crate::storage::persistence::VectorIndexCheckpointData {
>     self.get_default_vector_property_name()
>         .and_then(|property_name| {
>             let entry = self.vector_indexes.get(&property_name)?;
>             Some(VectorIndexCheckpointData::enabled(
>                 property_name,
>                 entry.value().config.clone(),
>             ))
>         })
>         .unwrap_or_else(VectorIndexCheckpointData::disabled)
> }
> ```
> 
> **Problem**: This only persists the **alphabetically first** index. If multiple vector indexes are configured, only one is saved to checkpoints, and the others are lost on recovery.
> 
> **Impact**:
> 
> * **Data loss risk**: Multi-property setups lose all but one index after checkpoint recovery
> * **Silent failure**: No warning that indexes are being dropped
> * **Violates expectations**: The code supports multi-property but doesn't persist it
> 
> **Recommendation**:
> 
> * Either update `VectorIndexCheckpointData` to support multiple properties (breaking change to checkpoint format)
> * Or add validation to reject multi-property setups if persistence is enabled
> * Or add prominent documentation warning about this limitation
> 
> #### **2. Potential Race Condition in `get_default_vector_property_name()`**
> **Location**: `src/storage/current.rs:1307-1319`
> 
> The method reads from `DashMap` multiple times:
> 
> ```rust
> fn get_default_vector_property_name(&self) -> Option<String> {
>     if self.vector_indexes.is_empty() {  // Read 1
>         return None;
>     }
>     if self.vector_indexes.len() == 1 {  // Read 2
>         return self.vector_indexes.iter().next().map(|r| r.key().clone()); // Read 3
>     }
>     self.vector_indexes.iter().map(|r| r.key().clone()).min() // Read 4
> }
> ```
> 
> **Problem**: Between checking `len() == 1` and calling `.iter().next()`, another thread could:
> 
> * Remove the only index → `.next()` returns `None` (correct behavior)
> * Add another index → `.next()` returns one of two (not alphabetically first)
> 
> **Impact**: Low severity (non-determinism only affects which property is selected in rare race), but violates the "deterministic" claim in docs.
> 
> **Recommendation**: Accept the race as benign (it's already concurrent) or document that determinism is "eventual" under concurrent modifications.
> 
> #### **3. Inconsistent Behavior: `vector_count()` Only Counts Default Property**
> **Location**: `src/storage/current.rs:2395-2400`
> 
> ```rust
> pub fn vector_count(&self) -> usize {
>     self.get_default_vector_property_name()
>         .and_then(|prop_name| self.vector_indexes.get(&prop_name))
>         .map(|entry| entry.value().index.len())
>         .unwrap_or(0)
> }
> ```
> 
> **Problem**: In a multi-property setup, this only returns the count for the **alphabetically first** property, not the total vector count across all indexes.
> 
> **Impact**:
> 
> * Misleading statistics for query planner
> * If the "default" property has few vectors but other properties have many, statistics are wrong
> * Query planner may make suboptimal decisions
> 
> **Recommendation**:
> 
> * Either document clearly that this is for the default property only
> * Or add a `total_vector_count()` method that sums across all properties
> * Or rename to `default_property_vector_count()` for clarity
> 
> ### 📋 **Code Quality**
> #### **Positive**:
> * ✅ Code follows Rust best practices
> * ✅ Proper use of `Option` and `Result` error handling
> * ✅ No `.unwrap()` in production code paths
> * ✅ Good inline documentation
> * ✅ Clippy-compliant (based on project standards)
> 
> #### **Minor Issues**:
> * **Optimization note**: The `min()` call on line 1319 requires collecting all keys. For large numbers of indexes (unlikely), consider caching the default property name.
> * **Naming consistency**: `get_indexed_property_name()` and `get_vector_property_name()` are aliases doing the same thing. Consider deprecating one for clarity.
> 
> ### 🧪 **Test Coverage**
> **New Tests**: Good additions, but missing edge cases:
> 
> 1. ✅ **Covered**: Deterministic selection with multiple properties
> 2. ✅ **Covered**: Vector count calculation
> 3. ❌ **Missing**: Checkpoint persistence with multiple properties (would expose Issue [Add Claude Code GitHub Workflow #1](https://github.com/madmax983/GallifreyDB/pull/1))
> 4. ❌ **Missing**: Concurrent index addition/removal during `get_default_vector_property_name()`
> 5. ❌ **Missing**: Behavior when all indexes are removed then one is re-added
> 
> **Recommendation**: Add a test that:
> 
> * Enables multiple vector indexes
> * Creates a checkpoint
> * Recovers from checkpoint
> * Verifies all indexes are restored (this will currently **fail**)
> 
> ### 🔒 **Security**
> ✅ No security concerns identified. The refactor doesn't change security boundaries or introduce new attack vectors.
> 
> ### ⚡ **Performance**
> **Improvements**:
> 
> * ✅ Removing `RwLock` eliminates lock contention on `vector_index_state`
> * ✅ Direct `DashMap` access is more efficient than state → DashMap indirection
> 
> **Concerns**:
> 
> * ⚠️ `get_default_vector_property_name()` with `.min()` is O(n) in number of properties. Not a problem for typical workloads (1-5 properties), but could be cached.
> * ⚠️ Multiple `DashMap` traversals in edge cases (see race condition above)
> 
> ### 📝 **Recommendations**
> #### **Must Fix Before Merge**:
> 1. **Checkpoint Persistence**: Resolve the multi-property checkpoint issue (see Issue [Add Claude Code GitHub Workflow #1](https://github.com/madmax983/GallifreyDB/pull/1) above). Either:
>    
>    * Update checkpoint format to support multi-property
>    * Add validation to prevent multi-property setups with persistence
>    * Document the limitation prominently in `enable_vector_index()` docs
> 
> #### **Should Fix**:
> 2. **Clarify `vector_count()` Behavior**: Rename or document that it's only for the default property
> 3. **Add Persistence Test**: Test multi-property checkpoint/recovery to validate fix for [Add Claude Code GitHub Workflow #1](https://github.com/madmax983/GallifreyDB/pull/1)
> 
> #### **Nice to Have**:
> 4. Document race condition behavior in `get_default_vector_property_name()`
> 5. Consider caching the default property name for performance (micro-optimization)
> 
> ### 🎯 **Verdict**
> **Overall Assessment**: Good refactoring that successfully removes legacy state, but has a **critical issue with checkpoint persistence** that could cause data loss in multi-property configurations.
> 
> **Recommendation**: **Request Changes** - the checkpoint persistence issue must be addressed before merging to prevent silent data loss.
> 
> Once the checkpoint issue is resolved, this is a solid cleanup that improves code maintainability and performance.
> 
> ### 📚 **References**
> * **Issue [Comprehensive Vector Index API: Multi-property support and unified DX #389](https://github.com/madmax983/GallifreyDB/issues/389)**: Multi-property vector indexes
> * **CLAUDE.md**: Testing requirements (85%+ coverage maintained ✅)
> * **CODING_STANDARDS.md**: Followed ✅
> * Related: `TemporalVectorIndexState` still exists and may benefit from similar cleanup in a future PR

@jules
